src/replication/DealReplicationWorker.ts

```diff
@@ -0,0 +1,327 @@
+            return;
+          }
+        } else {
+          let propose = `deal --http-url=${downloadUrl}`;
+          if (model.isOffline) {
+            propose = `offline-deal`;
+          }
+          dealCmd = `${boostCMD} ${propose} --provider=${miner}  --commp=${carFile.pieceCid} --car-size=${carFile.carSize} ` +
+            `--piece-size=${carFile.pieceSize} --payload-cid=${carFile.dataCid} --storage-price-per-epoch=${priceInAtto} ` +
+            `--verified=${model.isVerfied} --wallet=${model.client} --duration=${model.duration}`;
+        }
+
+        this.logger.debug(dealCmd);
+        let dealCid = 'unknown';
+        let errorMsg = '';
+        let state = 'proposed';
+        let retryCount = 0;
+        do {
+          try {
+            const cmdOut = await exec(dealCmd);
+            this.logger.info(cmdOut.stdout);
+            if (useLotus) {
+              if (cmdOut.stdout.startsWith('baf')) {
+                dealCid = cmdOut.stdout;
+                dealsMadePerSP++;
+                break; // success, break from while loop
+              } else {
+                errorMsg = cmdOut.stdout + cmdOut.stderr;
+                state = 'error';
+              }
+            } else {
+              const match = boostResultUUIDMatcher.exec(cmdOut.stdout);
+              if (match != null && match.length > 1) {
+                dealCid = match[1];
+                dealsMadePerSP++;
+                break; // success, break from while loop
+              } else {
+                errorMsg = cmdOut.stdout + cmdOut.stderr;
+                state = 'error';
+              }
+            }
+          } catch (error) {
+            this.logger.error('Deal making failed', error);
+            errorMsg = '' + error;
+            state = 'error';
+          }
+          this.logger.info('Waiting 30 seconds to retry');
+          // TODO expoential back off
+          await new Promise(resolve => setTimeout(resolve, 30000));
+          retryCount++;
+        } while (retryCount < 3);
+        await Datastore.DealStateModel.create({
+          client: model.client,
+          provider: miner,
+          dealCid: dealCid,
+          dataCid: carFile.dataCid,
+          pieceCid: carFile.pieceCid,
+          expiration: 0,
+          duration: model.duration,
+          price: priceInFil.toNumber(), // unit is Fil
+          verified: model.isVerfied,
+          state: state,
+          replicationRequestId: model.id,
+          datasetId: model.datasetId,
+          errorMessage: errorMsg
+        });
+        if (dealCid !== 'unknown') {
+          csvArray.push({
+            provider: miner,
+            deal_cid: dealCid,
+            filename: carFilename,
+            piece_cid: carFile.pieceCid,
+            url: downloadUrl
+          });
+        }
+      }
+      if (csvArray.length > 0) {
+        const csv = new ObjectsToCsv(csvArray);
+        const csvFilename = `/tmp/${miner}_${model.id}.csv`;
```

Comment:
What's the purpose of CSV?

Reply:
This is merely for my own purpose. I send this type of CSV to SP to let them import offline deals. I'm thinking maybe our SPs can benefit from this too.

Reply:
I think we need to properly design that. i.e. If the dataset is 100TiB , are you going to send out all deals first and then have a csv for all 3200 deals? - I think in this case, you can use a external script to extract the data from MongoDb and construct the csv, so that you can control the time range, the directory to save the csv and what data to get